Silabs EFR32 codelab first draft to publish for openthread.io

silabs-mithil opened the pull request:
Add EFR32 codelab to openthread.io
Please validate the preview on openthread.io and let me know of any issues or errors in rendering.

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+
+
+## Setup the Raspberry Pi as a Border Router
```
"Set up" when using it as a verb

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+2. You can either SSH onto the Raspberry Pi or choose to work with the Raspbian Desktop directly. Open a terminal.
+3. Make sure to update your local repositories and package manager (**apt-get update** and **apt-get upgrade** prior to installing Docker).
+
+NOTE: Make sure to reboot the Raspberry Pi after any updates.
```
> Note: Make sure....

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+*  [Supported Platforms](https://openthread.io/platforms/)
+    — discover all the platforms that support OpenThread
+    
+*  [Build OpenThread](https://openthread.io/guides/build/)
```
For files that are in this repo, we're using relative links now

../../guides/build/index.md

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+*  [Build OpenThread](https://openthread.io/guides/build/)
+    — further details on building and configuring OpenThread
+    
+*  [Thread Primer](https://openthread.io/guides/thread-primer/)
```
../../guides/thread-primer/index.md

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+## Further reading
+
+Check out  [openthread.io](https://openthread.io) and [GitHub](https://github.com/openthread) for a variety of OpenThread resources, including:
```
Can you add a / at the end of the link, needed for our import process

https://openthread.io/

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+<img src="img/Silabs/26b7f4f6b3ea0700.png" alt="26b7f4f6b3ea0700.png" width="624.00" />
+
+Google's [OpenThread](https://openthread.io) (OT) is an open-source implementation of Thread. Google has released OpenThread to make the networking technology used in Google Nest products more broadly available to developers, in order to accelerate the development of products for the connected home and commercial buildings. With a narrow platform abstraction layer and a small memory footprint, Op ...
```
Add a / at the end of the link for our import process

https://openthread.io/

Vyrastas replied:
This was not done

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+ */
+
+---
+id: openthread-hardware
```
id: silabs-openthread-hardware

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+# Build a Thread network with Silicon Labs EFR32 boards and OpenThread using Simplicity Studio v5
+
+[Codelab Feedback](https://github.com/openthread/openthread/issues)
```
https://github.com/openthread/ot-docs/issues

Vyrastas replied:
Link still points to openthread rather than ot-docs

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+ ![](img/Silabs/BRD4166A.jpeg)![BRD4168A](./img/Silabs/BRD4168A.jpeg)
+
+> You can update the GSDK or toolchain from Studio -> Help -> Update Software -> SDKs/Toolchains
```
> Note: You can update...

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+### Firmware setup
+
+> There are two ways to get started. Both options allow you to flash the firmwares that are required for this codelab.
```
> Note: There are...

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+We will create two projects. The ot-rcp project for BRD4166A and the ot-cli-ftd project for the two BRD4168A. Please follow these steps and select the appropriate sample application for your board.
+
+1. Open Studio's File menu and select New > Silicon Labs Project Wizard. The Target, SDK, and Toolchain Selection dialog opens. Do not change the default Simplicity IDE / GNU toolchain supported by OpenThread. Click Next.
```
For all buttons and menu options, use bold:

**New** > **Silicon Labs Project Wizard**

Please update elsewhere names of controls like this are used.

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+1. Open Studio's File menu and select New > Silicon Labs Project Wizard. The Target, SDK, and Toolchain Selection dialog opens. Do not change the default Simplicity IDE / GNU toolchain supported by OpenThread. Click Next.
+
+   Target boards: Shows the selected radio board (BRD4168A) along with the main board (BRD4001A)
+
+   Target Device: This field shows the microcontroller chip (MCU) onboard. The BRD4168A has an EFR32MG13 MCU onboard.
+
+   SDK: Here you can select the SDK version of OT that you are working with. For this codelab we select OpenThread 1.1.2.0
+
+   IDE/ Toolchain: The toolchain that will be used for compiling the OT project. We use GNU ARM 7.2.1
```
These should be formatted as an indented list under the 1. bullet:

* Target boards: Shows the selected radio board (BRD4168A) along with the main board (BRD4001A)
* Target Device: This field shows the microcontroller chip (MCU) onboard. The BRD4168A has an EFR32MG13 MCU onboard.
* SDK: Here you can select the SDK version of OT that you are working with. For this codelab we select OpenThread 1.1.2.0
* IDE/ Toolchain: The toolchain that will be used for compiling the OT project. We use GNU ARM 7.

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+   IDE/ Toolchain: The toolchain that will be used for compiling the OT project. We use GNU ARM 7.2.1
+
+   > The IAR Embedded Workbench for ARM (IAR EWARM) Compiler is currently not supported with OpenThread.
```
Do not indent and format as a note

> Note: The IAR...

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+ The response indicates the payload received and that the communication is successful. Repeat the process to ping FTDs from the OTBR.
+
+**You've created a Thread network!**
```
Let's add a level 2 Congratulations! header here, and have that be the last step.

Further Reading can be a level 3 section under it.

silabs-mithil commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+
+
+
+## Getting started
```
> This step is quite lengthy, I suggest we break it up into the following so it's more digestible:
> 
> Hardware setup
> Firmware setup - demos
> Firmware setup - projects
> 
> Ideally in a codelab the user shouldn't jump around or skip steps, though, so it would be better to just have one setup method. Would you be willing to just have one method of firmware setup and remove the other, leaving it as an exercise for the reader?

@Vyrastas  I have reorganised the secion into Hardware and Firmware setup.  Firmware setup - demos is put as an optional exercise in the end. Is it possible to link this optional section in the earlier step under Firmware set up point 2 Demos?

silabs-mithil commented:
@Vyrastas Thank you for all the suggestions. I have revised the draft to incorporate them.

silabs-mithil commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,438 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
```
@Vyrastas , I have moved the license. It says the copyright belongs to the OpenThread Authors. I am trying to understand if it applies for this codelab or should I rephrase it to something like the copyright belongs to Silicon Labs?

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,461 @@
+
+Hardware:
+
+1. 3 EFR32MGxx radio boards - you can use any combination of these devices. I used a BRD4166A as the RCP and two BRD4168As as Full Thread Devices.
```
Don't refer to yourself in the text, use something like "this codelab uses a BRD4166A..."

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,461 @@
+
+  * Gecko SDK Suite 3.1.0 or later and Silicon Labs OpenThread SDK.
+
+    > Note: The IAR Embedded Workbench for ARM (IAR EWARM) Compiler is currently not supported with OpenThread.
```
Add this before the note, don't indent this note.

<noparse>> aside positive</noparse>
<noparse>></noparse>

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,461 @@
+
+ ![](img/BRD4166A.jpeg)![BRD4168A](./img/BRD4168A.jpeg)
+
+> Note: You can update the GSDK or toolchain from **Studio -> Help -> Update Software -> SDKs/Toolchains**
```
Add this before the note

<noparse>> aside positive</noparse>
<noparse>></noparse>

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,461 @@
+
+After completing this section, your Thread network hardware set up will look like this.
+
+> Not shown in the image is the WSTKs connected to a Windows/Linux/Mac Host system via USB and the Raspberry Pi 3B+ is connected to the internet using an Ethernet cable. 
```
Add this before the note

<noparse>> aside positive</noparse>
<noparse>></noparse>

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,461 @@
+
+To launch the Console interface, in the Simplicity IDE perspective **right-click on your J-Link device** in the Devices View / Debug Adapters Window. Choose **Launch Console**. To get a prompt on the Console, choose **Serial 1** tab and **press Enter**. Check the state of the FTD node.
+
+> Note: For convenience, I also right clicked and renamed the debug adapters as `ot-cli-ftd 1` and `ot-cli-ftd 2`. 
```
Add this before the note

<noparse>> aside positive</noparse>
<noparse>></noparse>

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,461 @@
+2. You can either SSH onto the Raspberry Pi or choose to work with the Raspbian Desktop directly. Open a terminal.
+3. Make sure to update your local repositories and package manager (**apt-get update** and **apt-get upgrade** prior to installing Docker).
+
+> Note: Make sure to reboot the Raspberry Pi after any updates.
```
Add this before the note

<noparse>> aside positive</noparse>
<noparse>></noparse>

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,461 @@
+|       |                                                      | Wait 20 seconds while the device joins and configures itself. |                   |
+| 6     | `state`                                              | View network configuration.                                  | child<br/>Done    |
+
+> Note: Because of the self configuring nature of a Thread network and these being Full Thread devices, either or both the FTDs may eventually become  routers. This can be found out using the `state` command.
```
Add this before the note

<noparse>> aside positive</noparse>
<noparse>></noparse>

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,461 @@
+
+5. BRD4304a
+
+   This list may be updated in the future SDK releases to include more radio boards. Please refer to the Silicon Labs OpenThread SDK release notes under Documentation for a full list of supported parts.
```
Don't indent this line, the parser thinks it's part of the list.

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,461 @@
+
+Flash the following demos on the respective boards. To flash, select your board under Debug Adapters on the left and click on RUN for the corresponding sample applications. A pop up window will show the flash progress.
+
+1. BRD4166A: **ot-rcp**
+
+   This device will function as a Radio Co Processor to the OT Border Router. We will use this device to create a Thread network and onboard the other two devices on the Thread network. Being a Border Router this device also acts as a gateway for the devices in the Thread network to communicate over the internet.
```
Can you have the text as part of the list item rather than separated?  So like

1. BRD4166A: **ot-rcp** — This device will...

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,461 @@
+
+   This device will function as a Radio Co Processor to the OT Border Router. We will use this device to create a Thread network and onboard the other two devices on the Thread network. Being a Border Router this device also acts as a gateway for the devices in the Thread network to communicate over the internet.
+
+2. Two BRD4168A: **ot-cli-ftd**
+
+   These two devices will act as Full Thread Devices. They will join the Thread network created by the OTBR.
```
Same thing here as above

Vyrastas reviewed (changes requested):
Thanks for all the updates.  A few more comments... most are from a fix to the note formatting, I gave you the wrong guidance before (sorry!).

I think we'll be good after this.

silabs-mithil commented:
@Vyrastas Thank you once again for all the suggestions. I have revised the draft to incorporate them. Also moved the [status] to publish and added a button to download Studio on line 83

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
```diff
@@ -0,0 +1,475 @@
+---
+id: silabs-openthread-hardware
+summary: In this Codelab, you'll program OpenThread on real hardware, create and manage a Thread network, and pass messages between nodes.
+status: [publish]
```
The status should be [final]

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
````diff
@@ -0,0 +1,475 @@
+
+```console
+> ipaddr
+fd5c:c6b:3a17:40b9:0:ff:fe00:fc00					# Leader Anycast Locator (ALOC)
+fd5c:c6b:3a17:40b9:0:ff:fe00:1800					# Routing Locator (RLOC)
+fd5c:c6b:3a17:40b9:84e2:bae8:bd5b:fa03		# Mesh-Local EID (ML-EID)
+fe80:0:0:0:c449:ca4a:101f:5d16						# Link-Local Address (LLA)
````
Can you align all the comments here to the Mesh-Local EID one so they don't wrap?

Vyrastas commented on site/en/codelabs/silabs-openthread-hardware/index.lab.md:
````diff
@@ -0,0 +1,475 @@
+From both FTDs ping the OTBR using the OTBR's RLOC address.
+
+```console
+ping fd5c:c6b:3a17:40b9:0:ff:fe00:1800
+Done
+> 
+> 16 bytes from fd5c:c6b:3a17:40b9:0:ff:fe00:1800: icmp_seq=3 hlim=64 time=30ms
+16 bytes from fd5c:c6b:3a17:40b9:0:ff:fe00:1800: icmp_seq=3 hlim=64 time=52ms
````
If this is using the OT CLI then ping should have the > prompt... can you double-check the output?  My understanding is that it would look like this, but I can't quickly test at the moment:

```
> ping fd5c:c6b:3a17:40b9:0:ff:fe00:1800
> 16 bytes from fd5c:c6b:3a17:40b9:0:ff:fe00:1800: icmp_seq=3 hlim=64 time=30ms
16 bytes from fd5c:c6b:3a17:40b9:0:ff:fe00:1800: icmp_seq=3 hlim=64 time=52ms
```

silabs-mithil replied:
Yes, > is present for the CLI

Vyrastas reviewed (changes requested):
Just a few more things!  After this we should be good to go.

silabs-mithil commented:
Revised to correct the last few errors

Vyrastas reviewed (approved):
LGTM, thanks!

jwhui reviewed (approved):
Looks great. Thanks! 👍🏼

Vyrastas commented:
@silabs-mithil I'll get this published on openthread.io tomorrow.

Vyrastas commented:
@silabs-mithil This has been published:

https://openthread.io/guides#hardware_codelabs
https://openthread.io/codelabs/silabs-openthread-hardware

If anything needs fixing, please do it here in GitHub and we'll import the update to openthread.io.  Thanks for the contribution!

silabs-mithil commented:
Looks good, thank you @Vyrastas and @jwhui
